### PR TITLE
Draft: Add bzlmod minimal support

### PR DESCRIPTION
**Pull request by mateuszkuta256**
### Description
This is the first step of migration to Bzlmod, see https://bazel.build/external/migration

I'm gonna create a few further pull requests to bring all the changes described here: https://github.com/bazelbuild/rules_scala/issues/1482#issuecomment-1907672400
The migration process will finish once WORKSPACE.bzlmod is empty

### Motivation
Resolve https://github.com/bazelbuild/rules_scala/issues/1482

### CODE REVIEW

File: scala/private/phases/phase_dependency.bzl [hunk 1]
```diff
@@ -85,11 +85,11 @@ def _get_unused_deps_mode(ctx):
 
 def _is_target_included(target, includes, excludes):
     for exclude in excludes:
-        if target.startswith(exclude):
+        if target.startswith(exclude) or target.startswith("@" + exclude):
```
Comment (mateuszkuta256) [inline scala/private/phases/phase_dependency.bzl:88]: https://bazel.build/external/overview#canonical-repo-name
for blzmod labels can be like "@@..."

File: scala/scala_maven_import_external.bzl [hunk 2]
```diff
@@ -64,7 +64,7 @@ def _jvm_import_external(repository_ctx):
     if (repository_ctx.attr.generated_linkable_rule_name and
         not repository_ctx.attr.neverlink):
         fail("Only use generated_linkable_rule_name if neverlink is set")
-    name = repository_ctx.attr.generated_rule_name or repository_ctx.name
+    name = repository_ctx.attr.repo_name or repository_ctx.attr.generated_rule_name or repository_ctx.name
```
Comment (mateuszkuta256) [inline scala/scala_maven_import_external.bzl:67]: names of repositories issued by bzlmod module extensions will look like:
`_main~non_module_deps~new_repo`
this is why I added explicit parameter to achieve: `@new_repo`

File: test/shell/test_scala_config.sh [hunk 3]
```diff
@@ -18,7 +18,7 @@ test_classpath_contains_2_13() {
 
 test_scala_config_content() {
   bazel build --repo_env=SCALA_VERSION=0.0.0 @io_bazel_rules_scala_config//:all 2> /dev/null
-  grep "SCALA_MAJOR_VERSION='0.0'" $(bazel info output_base)/external/io_bazel_rules_scala_config/config.bzl
+  grep "SCALA_MAJOR_VERSION='0.0'" $(bazel info output_base)/external/*/config.bzl
```
Comment (mateuszkuta256) [inline test/shell/test_scala_config.sh:21]: when built with bzlmod target will be in sth like:
`/external/~main_non_module_deps~io_bazel_rules_scala_config/config.bzl`

File: third_party/repositories/scala_2_11.bzl [hunk 4]
```diff
@@ -516,4 +516,16 @@ artifacts = {
         "artifact": "org.typelevel:kind-projector_%s:0.13.2" % scala_version,
         "sha256": "8f7287973f7f8fc9372b59d36120e3fac5839344f65c8f640351794e8907145c",
     },
+    "io_bazel_rules_scala_scala_interfaces": {
```
Comment (mateuszkuta256) [inline third_party/repositories/scala_2_11.bzl:519]: looks kinda odd, but I couldn't figure out a better solution
the thing is we have to specify all the possible repos in a MODULE.bazel file
there's no access to SCALA_VERSION property from there, so cannot do sth like "use this repo only if scala_version == 3"
I made a doc comment in scala/scala_maven_import_external.bzl repository def which is meant to be used only for bzlmod

**Comment by mateuszkuta256** [comment]
I turn into a draft because will surely cause conflicts with [these changes](https://github.com/bazelbuild/rules_scala/issues/1290#issuecomment-1912045651):
> SCALA_VERSION dependent code has to be refactored to read this information from the toolchain

**Comment by simuons** [comment]
Hi, @mateuszkuta256, thanks for taking this. So as I understand this is somewhat "blocked" by required changes to toolchains and `SCALA_VERSION`. I'll reply in [multi scala issue](https://github.com/bazelbuild/rules_scala/issues/1290#issuecomment-1912045651) shortly. This change and multi scala are the big ones. So maybe lets first solve multi scala and toolchains and then bzlmod (i have a hunch that proper toolchains would make bzlmod migration easier).

**Comment by simuons** [comment]
Closing in favour of https://github.com/bazelbuild/rules_scala/issues/1482

**Pull request closed without merging**